django/views/decorators/http.py

```diff
@@ -118,25 +133,39 @@ def get_last_modified():
             if response is None:
                 response = func(request, *args, **kwargs)
 
-            # Set relevant headers on the response if they don't already exist
-            # and if the request method is safe.
-            if request.method in ("GET", "HEAD"):
-                if res_last_modified and not response.has_header("Last-Modified"):
-                    response.headers["Last-Modified"] = http_date(res_last_modified)
-                if res_etag:
-                    response.headers.setdefault("ETag", res_etag)
+            _process_response(request, response, res_etag, res_last_modified)
+
+            return response
+
+        @wraps(func)
+        async def inner_async(request, *args, **kwargs):
+            res_etag = _get_res_etag(request, *args, **kwargs)
+            res_last_modified = _get_last_modified(request, *args, **kwargs)
+
+            response = get_conditional_response(
+                request,
+                etag=res_etag,
+                last_modified=res_last_modified,
+            )
+
+            if response is None:
+                response = await func(request, *args, **kwargs)
+
+            _process_response(request, response, res_etag, res_last_modified)
 
             return response
 
-        return inner
+        return inner_async if asyncio.iscoroutinefunction(func) else inner_sync
```

Comment:
This was the neatest I could get it. The reason I can't just use `sync_async_wrapper` is that we would calculate `res_etag` and `res_last_modified` and use them in `process_request`, but then would need to access them again in the `process_response` function. To do this we would need to either:

1. Have some sort of dict / object to store them against (all the ones I could think of felt black-magic-y to me)
2. Recalculate them during `process_reponse` (but since we allow the user to pass a function, we can't guarantee that it will be idempotent and we might trigger things twice - this felt like a bad gotcha)

This is what the code partially looks like, with the problem highlighted:

```python
    def decorator(func):
        def _get_res_etag(request, *args, **kwargs):
            # The value from etag_func() could be quoted or unquoted.
            res_etag = etag_func(request, *args, **kwargs) if etag_func else None
            res_etag = quote_etag(res_etag) if res_etag is not None else None
            return res_etag

        def _get_last_modified(request, *args, **kwargs):
            # Compute values (if any) for the requested resource.
            if last_modified_func:
                dt = last_modified_func(request, *args, **kwargs)
                if dt:
                    if not timezone.is_aware(dt):
                        dt = timezone.make_aware(dt, datetime.timezone.utc)
                    return int(dt.timestamp())

        def _process_request(request, *args, **kwargs):
            res_etag = _get_res_etag(request, *args, **kwargs)
            res_last_modified = _get_last_modified(request, *args, **kwargs)
            return get_conditional_response(
                request,
                etag=res_etag,
                last_modified=res_last_modified,
            )

        def _process_response(request, response, **decorator_kwargs):

            # --- WE DON'T HAVE res_last_modified OR res_etag HERE ---

            # Set relevant headers on the response if they don't already exist
            # and if the request method is safe.
            if request.method in ("GET", "HEAD"):
                if res_last_modified and not response.has_header("Last-Modified"):
                    response.headers["Last-Modified"] = http_date(res_last_modified)
                if res_etag:
                    response.headers.setdefault("ETag", res_etag)

        return sync_async_wrapper(
            func,
            process_request=_process_request,
            _process_response=_process_response,
        )

    return decorator
```